# scPipe-ATAC

[![Travis build status](https://travis-ci.org/shaniAmare/scPipe-ATAC.svg?branch=master)](https://travis-ci.org/shaniAmare/scPipe-ATAC)
[![Coverage
Status](https://codecov.io/gh/shaniAmare/scPipe-ATAC/branch/master/graph/badge.svg)](https://codecov.io/gh/shaniAmare/scPipe-ATAC)

**This repository for scPipe upgrades is not maintained anymore. There can be a lot of bugs if you try to run this version. The active version is avaialble at [scPipe Upgrades](https://github.com/shaniAmare/scPipe)**.

scPipe-ATAC is a R module that contains capabilities of pre-processing scATAC-Seq data through <a href=https://github.com/LuyiTian/scPipe> scPipe </a>. 
This package allows barcode demultiplexing, peak calling and quality control of raw sequencing data generated by multiple single-cell ATAC-Seq sequencing protocols including 
10X, scATAC-Seq, dscATAC-Seq, dsciATAC-Seq, sciATAC-Seq, plate-based ATAC-Seq and scHTS-Seq.

scPipe-ATAC produces a feature-barcode count matrix that is essential for downstream analysis along with a user-friendly HTML report that summarises data quality. 
These results can be used as input for downstream analyses including normalization, visualization and statistical testing, through a *Single Cell Experiment* object.

The plan is to integrate fully functional scPipe-ATAC functions with the existing scPipe package and allow routine maintenance to happen there.

* [20/07/2020] scPipe-ATAC submitted as a poster to BioC2020 virtual conference
* [12/07/2020] incorporate bedtools2 to generate feature matrix
* [28/06/2020] incorporate macs2 to call peaks
* [20/06/2020] extend sc_trim_barcode() to generate the CB:Z: column from the BAM file
* [02/06/2020] scPipe-ATAC demultiplexing step completed; 0.4 seconds to process a 6MB of paired-end fastq reads
* [13/05/2020] initiate scPipe-ATAC repository on GitHub 

## Installation

### From GitHub (Developmental version)

```{r}
install.packages("devtools")
devtools::install_github("shaniAmare/scPipe-ATAC")
```

## Getting started (minimal example)

Open RStudio

Load the project

```{r}
devtools::install()
devtools::load_all()
```

Install the packages from sc_atac_setup.R

-- However, these package installation will be automatically handled within the package later

### Demultiplexing -----------------

```{r}
r1      <- system.file(
  "data","testfastq_S1_L001_R1_001.fastq.gz",package="scPipe-ATAC") 
r2      <- system.file(
  "data","testfastq_S1_L001_R3_001.fastq.gz",package="scPipe-ATAC") 
barcode_fastq      <- system.file(
  "data","testfastq_S1_L001_R2_001.fastq.gz",package="scPipe-ATAC") 
Barcode_path <- system.file("data","barcode.csv",package="scPipe-ATAC")

sc_atac_trim_barcode (r1 = r1, r2 =  r2, bc_file =  barcode_fastq, output_folder = "")
```

### Aligning to reference -----------------

```{r}
reference <- system.file("data", "genome.fa")
r1 <- system.file("scPipe-atac-output", "demux_testfastq_S1_L001_R1_001.fastq.gz")
r2 <- system.file("scPipe-atac-output", "demux_testfastq_S1_L001_R3_001.fastq.gz")

sc_atac_aligning(ref = "data/genome.fa", readFile1 = r1, readFile2=r2)
```

### Tagging the aligned BAM file -----------------

```{r}
bam <- system.file("scPipe-atac-output", "demux_testfastq_S1_L001_R1_001_aligned.bam")

sc_atac_bam_tagging(inbam = bam, outbam="", bam_tags = list(bc="CB", mb="OX"), nthreads=1)
```

### Feature counting -----------------

```{r}
sorted_bam <- system.file("scPipe-atac-output", "testfastq_S1_L001_R1_001_tagged_sorted.BAM") # here is the issue the BAM file extension should be .bam and not .BAM
features   <- system.file("scPipe-atac-output", "NA_peaks.narrowPeak")

sc_atac_feature_counting(insortedbam = sorted_bam, feature_input = features, bam_tags = list(bc="CB", mb="OX"),approach = "peak")
```

### Currently ends here...

## Data Preprocessing

To be completed ...

## Acknowledgments

To be completed ...
